src/libstd/sync/once.rs

```diff
@@ -347,122 +388,127 @@ impl Once {
-                // not RUNNING.
                 _ => {
-                    assert!(state & STATE_MASK == RUNNING);
-                    let mut node = Waiter {
-                        thread: Some(thread::current()),
-                        signaled: AtomicBool::new(false),
-                        next: ptr::null_mut(),
-                    };
-                    let me = &mut node as *mut Waiter as usize;
-                    assert!(me & STATE_MASK == 0);
-
-                    while state & STATE_MASK == RUNNING {
-                        node.next = (state & !STATE_MASK) as *mut Waiter;
-                        let old = self.state.compare_and_swap(state,
-                                                              me | RUNNING,
-                                                              Ordering::SeqCst);
-                        if old != state {
-                            state = old;
-                            continue
-                        }
-
-                        // Once we've enqueued ourselves, wait in a loop.
-                        // Afterwards reload the state and continue with what we
-                        // were doing from before.
-                        while !node.signaled.load(Ordering::SeqCst) {
-                            thread::park();
-                        }
-                        state = self.state.load(Ordering::SeqCst);
-                        continue 'outer
-                    }
+                    // All other values must be RUNNING with possibly a
+                    // pointer to the waiter queue in the more significant bits.
+                    assert!(state_and_queue & STATE_MASK == RUNNING);
+                    wait(&self.state_and_queue, state_and_queue);
+                    state_and_queue = self.state_and_queue.load(Ordering::Acquire);
                 }
             }
         }
     }
 }
 
+fn wait(state_and_queue: &AtomicUsize, current_state: usize) {
+    // Create the node for our current thread that we are going to try to slot
+    // in at the head of the linked list.
+    let mut node = Waiter {
+        thread: Cell::new(Some(thread::current())),
+        signaled: AtomicBool::new(false),
+        next: ptr::null(),
+    };
+    let me = &node as *const Waiter as usize;
+    assert!(me & STATE_MASK == 0); // We assume pointers have 2 free bits that
+                                   // we can use for state.
+
+    // Try to slide in the node at the head of the linked list.
+    // Run in a loop where we make sure the status is still RUNNING, and that
+    // another thread did not just replace the head of the linked list.
+    let mut old_head_and_status = current_state;
+    loop {
+        if old_head_and_status & STATE_MASK != RUNNING {
+            return; // No need anymore to enqueue ourselves.
+        }
+
+        node.next = (old_head_and_status & !STATE_MASK) as *const Waiter;
```

Comment:
This is where you still write to `node.next`, which is as bad as doing `&mut node.next` -- it says that this is unique, meaning that future reads through `me` become illegal.

Reply:
You are right! I'll think about it a little, hopefully some reordering is also an alternative... And otherwise raw pointers all around, but that makes me less comfortable.

Reply:
How can this be problematic? The mutable reference ist unique, since the next field is only written to *before* the node becomes visible to other threads. Isn't this like mutating some value and *then* sending a shared reference to some other thread?

Reply:
You can't just take a random `usize` to the right address and use that; the corresponding address has to have an item in its borrow stack indicating that usize-ptrs are allowed to access it. This item gets added when you cast a reference to a raw pointer, but here it gets *removed* again!

When you use a mutable reference, that means "no other pointer is allowed to point here". This includes pointers "hidden" in integers. Maybe a better way to understand it is to think of this as: "all future accesses to this location must happen with pointers that are derived from me in the future". No pointer ever gets derived again from `node`, and thus no pointer ever again acquires the permission to access `node.next`.

Reply:
Wat, is any of this documented somewhere?

So `me` is a raw `*const` derived from a shared reference and instantly turned into a `usize`. Then it is mutated (using a `&mut`) once or a bunch of times and then another shared reference is created to read from the atomic `signalled` variable, wouldn't that mean that there *is* a shared reference in this borrow stack?

Reply:
> Wat, is any of this documented somewhere?

The normative docs are "don't violate the aliasing rules". Whatever that means.

My proposed model is documented at https://github.com/rust-lang/unsafe-code-guidelines/blob/master/wip/stacked-borrows.md.

> So me is a raw *const derived from a shared reference and instantly turned into a usize. Then node it is mutated (using a &mut) once or a bunch of times and then another shared reference is created to read from the atomic signalled variable, wouldn't that mean that there is a shared reference in this borrow stack?

There's a shared reference on the borrow stack but no item permitting raw integers any access. Shared references are not a wildcard; they still carefully control that only currently valid shared references may be used. (See https://github.com/rust-lang/unsafe-code-guidelines/issues/87 for what happens when we are too relayed about shared references. It's not good.)

We have no clear idea yet what the interact between integer-ptr casts and provenance is. C and LLVM are suffering from [miscompilations](https://bugs.llvm.org/show_bug.cgi?id=35229) due to ignoring these issues.